### ISSUE TITLE
Mesh generation algorithm

### ISSUE THREAD

**Issue by 0xCoto** [BlitPlatform/compute-servers#2]
Build a reliable tool to properly mesh CAD models, in order to allow the solver to reliably interpret and handle the given geometry with ease Since the solver is based on the finite-difference time-domain method, the algorithm should obviously produce a rectilinear mesh (consisting of hexahedral grid cells).

**Comment by thomaslepoix** [comment]
Here is the dedicated project: https://github.com/Open-RFlab/openemsh

**Comment by 0xCoto** [comment]
Alright, I think I've got something very promising... My algorithm automatically applies local refinement to edges, global refinement obeying the λ/15 rule (fraction adjustable). This was generated fully automatically for a patch STL, and I couldn't be too much happier - the total number of cells is **72x62x56 = 249,984**, which is fantastic:

![image](https://user-images.githubusercontent.com/25392776/171784983-8643e9ff-17a1-4d30-914c-0c98688b977d.png)

I've also made it so it takes into account the absorbing boundary condition for PML_x, for any `x` number of extra cells. I've currently set `max_cell` (global setting; derived from λ/15) for the padding size of the padding cells:.
![image](https://user-images.githubusercontent.com/25392776/171784980-35ce64f6-b989-4dc5-9ff8-e998e1f39918.png)

**Comment by 0xCoto** [comment]
Revised the code a bit to make the mesh smoother. The points are now spread out in a more uniform/equidistant manner, and the mesh looks amazing (with fewer elements at **76x65x38=187,720**):

![image](https://user-images.githubusercontent.com/25392776/171933344-17d9ff88-5064-4f4e-95ba-a97f5ac83279.png)
![image](https://user-images.githubusercontent.com/25392776/171933350-7b53d88e-b984-4e60-a89e-ccaf9bbd0c2d.png)

Also, the generation of the mesh is extremely efficient, yielding results in about **0.5s**!

**Comment by 0xCoto** [comment]
To-do: port refinement

**Comment by thomaslepoix** [comment]
Did you try to compare the simulation results of a such meshed model with the measures of the real piece or with a canonical handmade mesh? because what I see here does not comply with the [principles](https://openems.de/index.php/FDTD_Mesh.html) detailed in the documentation (thirds rule and *2 maximum width factor between two adjacent intervals). That being said I don't really know to which extend compliance to this principles helps the simulation accuracy.

**Comment by 0xCoto** [comment]
@thomaslepoix Yup, I definitely haven't forgotten about the thirds rule and 2x factor. I'm still tracing the origins of these rules, because they don't seem to be applicable to the finite integration technique (FIT) which works in a similar manner to FDTD. I've found some info about the thirds rule, but still searching for the 2x factor. Whether these are really necessary or not for fine meshes is not yet determined, but I'm trying to run some tests.

---

On that note, I've got two XML files: one created with Octave, and the other one created by my Python script. Although they look identical in terms of their content (and AppCSXCAD renders both models/meshes OK), it seems like the Python-generated XML with my automatic mesh algorithm (see [**`fails_python.zip`**](https://github.com/BlitPlatform/compute-servers/files/8872239/fails_python.zip)) gets stuck at `Running FDTD engine` with no following output:
```
Create FDTD operator (compressed SSE + multi-threading)
Operator::CalcGaussianPulsExcitation: Requested excitation pusle would be 255489276 timesteps or 5.72958e-009 s long. Cutting to max number of timesteps!
FDTD simulation size: 83x68x46 --> 259624 FDTD cells
FDTD timestep is: 2.24259e-017 s; Nyquist rate: 11147822 timesteps @2e+009 Hz
openEMS::SetupFDTD: Warning, the timestep seems to be very small --> long simulation. Check your mesh!?
Excitation signal length is: 60000 timesteps (1.34555e-012s)
Max. number of timesteps: 60000 ( --> 1 * Excitation signal length)
openEMS::SetupFDTD: Warning, max. number of timesteps is smaller than three times the excitation.
        You may want to choose a higher number of max. timesteps...
Create FDTD engine (compressed SSE + multi-threading)
Running FDTD engine... this may take a while... grab a cup of coffee?!?
```

Whereas, the Octave-generated XML (see [**`runs_octave.zip`**](https://github.com/BlitPlatform/compute-servers/files/8872241/runs_octave.zip)) seems to proceed fine:
```
Create FDTD operator (compressed SSE + multi-threading)
FDTD simulation size: 117x119x118 --> 1.64291e+006 FDTD cells
FDTD timestep is: 7.79191e-014 s; Nyquist rate: 1604 timesteps @4.00057e+009 Hz
openEMS::SetupFDTD: Warning, the timestep seems to be very small --> long simulation. Check your mesh!?
Excitation signal length is: 12256 timesteps (9.54977e-010s)
Max. number of timesteps: 30000 ( --> 2.44778 * Excitation signal length)
openEMS::SetupFDTD: Warning, max. number of timesteps is smaller than three times the excitation.
        You may want to choose a higher number of max. timesteps...
Create FDTD engine (compressed SSE + multi-threading)
Running FDTD engine... this may take a while... grab a cup of coffee?!?
[@        8s] Timestep:          401 || Speed:   74.3 MC/s (2.210e-002 s/TS) || Energy: ~2.23e-020 (- 0.00dB)
[@       17s] Timestep:          802 || Speed:   74.8 MC/s (2.195e-002 s/TS) || Energy: ~2.01e-019 (- 0.00dB)
[@       26s] Timestep:         1203 || Speed:   74.8 MC/s (2.198e-002 s/TS) || Energy: ~1.25e-018 (- 0.00dB)
[@       35s] Timestep:         1604 || Speed:   74.0 MC/s (2.219e-002 s/TS) || Energy: ~5.92e-018 (- 0.00dB)
...
```

Not sure what the problem is, but if you happen to have any ideas why one XML appears to run and the other one doesn't, that would greatly help me proceed to run some tests to evaluate the quality of the mesh.

**Comment by 0xCoto** [comment]
Just parametrized everything in the code so the mesh is easily customizable. Latest refinement looks closer to FIT's golden standard default:

![image](https://user-images.githubusercontent.com/25392776/172966255-b274128d-cc2b-4250-9b83-4ab1b4230b65.png)

I've also added a proof of concept ratio refinement ( ≲2x rule for neighboring cells) at [L167-L178](https://github.com/BlitPlatform/compute-servers/blob/main/patch.py#L167-L178), but it's still experimental and I didn't spend much time improving it until I can run some tests to ensure whether this rule is necessary or not.

Same goes for thirds rule: I can probably implement it (might require a bit more work if I do material-based refinement), but I want to run some accuracy tests first.

---

Another important milestone (experimental, but a proof of concept nontheless) is automatic mesh generation of curved, non-planar models (with the help of automatic tetrahedral mesh generation with `gmsh` as a starting point):

![image](https://user-images.githubusercontent.com/25392776/172967296-02a3a178-0d1e-4a45-bbce-d5b04f3494c6.png)

All automatic! 🙂

**Comment by 0xCoto** [comment]
Made some small improvements so grid lines introduced from edge refinements and what not don't end up right next to each other (83x71x83=200,362 cells).
![image](https://user-images.githubusercontent.com/25392776/177894595-4d8f7844-f363-4348-8740-c532072d31e7.png)

Side-view:
![image](https://user-images.githubusercontent.com/25392776/177895180-0cbb9807-06ea-4468-86d3-0af9adc55bc0.png)

**Comment by 0xCoto** [comment]
To-do: I'll also look into Numba in case I can bring the mesh generation down to milliseconds (not that of an important optimization, but I think it'll be an easy implementation).
